cpp/src/arrow/filesystem/gcsfs_test.cc

```diff
@@ -105,9 +106,21 @@ class GcsTestbench : public ::testing::Environment {
         error += " (exe not found)";
         continue;
       }
-
-      server_process_ = bp::child(exe_path, "-m", "testbench", "--port", port_, group_);
-      if (server_process_.valid() && server_process_.running()) break;
+      // TODO: should I be worried about lifetime of this variable?
+      bp::ipstream output;
+
+      server_process_ = bp::child(exe_path, "-m", "testbench", "--port", port_, group_,
+                                  bp::std_err > output);
+      // Wait for message: "* Restarting with"
+      auto testbench_is_running = [&output](bp::child& process) {
+        std::string line;
+        while (process.valid() && process.running() && std::getline(output, line)) {
+          if (line.find("* Restarting with") != std::string::npos) return true;
+        }
+        return false;
+      };
+
+      if (testbench_is_running(server_process_)) break;
```

Comment:
C++ question: is there an easy way to create a task with a timeout? As is, this could hang if the process keeps running without sending the correct message.

Reply:
In the `while` above you could just record the start time and add a time check to the loop guard

Reply:
I think the case I'm wary of is that the program doesn't write anything to std err for a while. IIUC `get_line` blocks until a line is retrieved, so in that case it could hang.

Reply:
I suppose you could spawn a thread to do it, detach it, and synchronize manually; if too much time elapses, continue on

Reply:
It depends on the value of "easy".  I have not tested this, but allegedly you can use asynchronous I/O with `Boost.Process`:

https://www.boost.org/doc/libs/1_79_0/doc/html/process/reference.html#header.boost.process.io_hpp

The example where they use a future could be changed with a `wait_for()`:

```cc
boost::asio::io_context ios;
std::future<std::string> output;
system("ls", std_out > output, ios);

auto status = fut.wait_for(std::chrono::seconds(10));
if (status == std::future_status::timeout) { /* uh oh */ }
```

I expect you would need a thread or something to run with the `boost::asio::io_context`, something like:

```cc
  boost::asio::io_context ios;
  std::thead svc{[&] { ios.run_for(std::chrono::seconds(..)); }};
...
  // insert the code from above
  svc.join();
....

Reply:
I tried writing a synchronous version that used `peek()` before calling `getline()`.

Reply:
I guess the synchronous version should be fine for our purposes. The testbench is written in Python and Python's stderr is always line-buffered by default, even if redirected.